stat/roc.go

```diff
@@ -98,14 +98,22 @@ func ROC(cutoffs, y []float64, classes []bool, weights []float64) (tpr, fpr, thr
 		}
 		nPos += posWeight
 		nNeg += negWeight
-		if y[i] <= cutoffs[bin] {
+		// Count false negatives (in tpr) and true negatives (in fpr).
+		if y[i] < cutoffs[bin] {
 			tpr[bin] += posWeight
 			fpr[bin] += negWeight
 		}
 	}
+	lastTpr := tpr[bin]
+	lastFpr := fpr[bin]
+	for bin++; bin < len(cutoffs); bin++ {
+		tpr[bin] = lastTpr
+		fpr[bin] = lastFpr
+	}
 
 	invNeg := 1 / nNeg
 	invPos := 1 / nPos
+	// Convert negative counts to TPR and FPR.
 	for i := range tpr {
```

Comment:
Out of curiosity (and laziness), does [my approach](https://github.com/gonum/gonum/pull/1558/files#diff-07d8944b2b9171f3dfbb983bfc016c30c93b829e6d2b2c4c91b21de24f3c8dccR109) used in #1558 work? It's simpler than this and does less work.

Reply:
I could be wrong, but I think the case for which it wouldn't work is when there are more cutoffs above the maximum y value, I added a test case for this case (test case 26). I agree your solution is prettier though.

Reply:
Just tested with your new cases and yes, this simpler approach is correct.

Reply:
Huh, yes I see what you mean now. This is confusing, but I believe it is actually equivalent. How about I incorporate you're simpler approach, but also add a comment to try and explain what it's doing?

Reply:
SGTM. I would say,

```
// Convert negative counts to TPR and FPR for non-zero bins.
```

Note that the reason this works is that these bins are zero in the final result, but `make` has already initialised the elements to be zero. So we just work on bins that are non-zero and use the existing zeros for the final result.

Reply:
Yeah, cool. It's always an interesting tradeoff between efficiency and clarity. I suppose that's the point of comments. You happy for me to squash all these commit's together now?

Reply:
Yes, please. Note that this is not strictly a reversion. I'd say "stat: fix ROC cutoffs".

Do you want/need to add UWA to AUTHORS (separate commit)?

Reply:
Sounds good. 

Hmm, no need to add UWA but maybe I should change the contact to my personal gmail, because I don't have access to the adelaide email address anymore... good point.

Reply:
OK. Make that a new PR. It would just be adding your personal email address to the lines in A+C.